solution.py

```diff
@@ -1,83 +1,76 @@
-     
-        # Start the ROSAgent, which handles publishing images and subscribing to action 
+
+        # Start the ROSAgent, which handles publishing images and subscribing to action
         self.agent = ROSAgent()
 
-    def init(self, context, data):
-        context.info('init()')
+    def init(self, context: Context):
+        context.info("init()")
 
-    def on_received_seed(self, context, data):
+    def on_received_seed(self, context: Context, data: int):
         np.random.seed(data)
 
-    def on_received_episode_start(self, context, data):
-        context.info('Starting episode %s.' % data)
+    def on_received_episode_start(self, context: Context, data: EpisodeStart):
+        context.info("Starting episode %s." % data)
 
-    def on_received_observations(self, context, data):
-        jpg_data = data['camera']['jpg_data']
+    def on_received_observations(self, context: Context, data: Duckiebot1Observations):
+        jpg_data = data.camera.jpg_data
         obs = jpg2rgb(jpg_data)
         self.agent._publish_img(obs)
         self.agent._publish_info()
 
-    def on_received_get_commands(self, context, data):
+    def on_received_get_commands(self, context: Context, data: GetCommands):
+        # TODO: let's use a queue here. Performance suffers otherwise.
+        # What you should do is: *get the last command*, if available
+        # otherwise, wait for one command.
         while not self.agent.updated:
             time.sleep(0.01)
-            
+
         pwm_left, pwm_right = self.agent.action
         self.agent.updated = False
 
-        rgb = {'r': 0.5, 'g': 0.5, 'b': 0.5}
-        commands = {
-            'wheels': {
-                'motor_left': pwm_left,
-                'motor_right': pwm_right
-            },
-            'LEDS': {
-                'center': rgb,
-                'front_left': rgb,
-                'front_right': rgb,
-                'back_left': rgb,
-                'back_right': rgb
-
-            }
-        }
+        grey = RGB(0.5, 0.5, 0.5)
+        led_commands = LEDSCommands(grey, grey, grey, grey, grey)
+        pwm_commands = PWMCommands(motor_left=pwm_left, motor_right=pwm_right)
+        commands = Duckiebot1Commands(pwm_commands, led_commands)
 
-        context.write('commands', commands)
+        context.write("commands", commands)
 
     def finish(self, context):
-        context.info('finish()')
+        context.info("finish()")
 
 
 def jpg2rgb(image_data):
     """ Reads JPG bytes as RGB"""
-    from PIL import Image
-    import io
     im = Image.open(io.BytesIO(image_data))
-    im = im.convert('RGB')
-    data = np.array(im) 
+    im = im.convert("RGB")
+    data = np.array(im)
     assert data.ndim == 3
     assert data.dtype == np.uint8
     return data
 
 
-if __name__ == '__main__':
+if __name__ == "__main__":
     agent = ROSTemplateAgent()
-    wrap_direct(agent)
\ No newline at end of file
+    wrap_direct(agent, protocol_agent_duckiebot1)
```

Comment:
@CourchesneA protocol_agent_duckiebot1 was the right one to choose - `protocol_agent` gave you the protocol errors